app/controllers/user/sessions_controller.rb

```diff
@@ -0,0 +1,47 @@
+class User::SessionsController < Devise::SessionsController
+  def create
+    if session.has_key?(:user_sign_in_uid)
+      self.resource = User.find(session[:user_sign_in_uid])
+      session.delete(:user_sign_in_uid)
+    else
+      self.resource = warden.authenticate!(auth_options)
+    end
+
+    if resource.active_for_authentication? && resource.otp_module_enabled?
+      if params[:user][:otp_attempt].blank?
+        session[:user_sign_in_uid] = resource.id
+        sign_out(resource)
+        redirect_to user_two_factor_entry_url
```

Comment:
No need for this workaround with the second page, just lock Warden for this request after the `sign_out` so `current_user` and the likes do not sign the user in again. (this is due to `current_user` calling `warden.authenticate`, which performs a sign in for the user again due to the parameters for the database authenticatable stuff being present)

```suggestion
        warden.lock!
        render 'auth/two_factor_authentication'
```

--> then get rid of `def two_factor_entry` and the route